server/apps/immich/src/api-v1/user/user.core.ts

```diff
@@ -0,0 +1,155 @@
+import { IUserRepository, UserListFilter } from './user-repository';
+
+export class UserCore {
+  constructor(private userRepository: IUserRepository) {}
+
+  private async generateSalt(): Promise<string> {
+    return genSalt();
+  }
+
+  private async hashPassword(password: string, salt: string): Promise<string> {
+    return hash(password, salt);
+  }
+
+  async updateUser(authUser: AuthUserDto, userToUpdate: UserEntity, data: Partial<UserEntity>): Promise<UserEntity> {
+    if (!authUser.isAdmin && (authUser.id !== userToUpdate.id || userToUpdate.id != data.id)) {
+      throw new ForbiddenException('You are not allowed to update this user');
+    }
+
+    try {
+      const payload: Partial<UserEntity> = { ...data };
+      if (payload.password) {
+        const salt = await this.generateSalt();
+        payload.salt = salt;
+        payload.password = await this.hashPassword(payload.password, salt);
+      }
+      return this.userRepository.update(userToUpdate.id, payload);
+    } catch (e) {
+      Logger.error(e, 'Failed to update user info');
+      throw new InternalServerErrorException('Failed to update user info');
+    }
+  }
+
+  async createUser(createUserDto: CreateUserDto | CreateAdminDto | CreateUserOAuthDto): Promise<UserEntity> {
+    const user = await this.userRepository.getByEmail(createUserDto.email);
+    if (user) {
+      throw new BadRequestException('User exists');
+    }
+
+    if (!(createUserDto as CreateAdminDto).isAdmin) {
+      const localAdmin = await this.userRepository.getAdmin();
+      if (!localAdmin) {
+        throw new BadRequestException('The first registered account must the administrator.');
+      }
+    }
+
+    try {
+      const payload: Partial<UserEntity> = { ...createUserDto };
+      if (payload.password) {
+        const salt = await this.generateSalt();
+        payload.salt = salt;
+        payload.password = await this.hashPassword(payload.password, salt);
+      }
+      return this.userRepository.create(payload);
+    } catch (e) {
+      Logger.error(e, 'Create new user');
+      throw new InternalServerErrorException('Failed to register new user');
+    }
+  }
+
+  async get(userId: string, withDeleted?: boolean): Promise<UserEntity | null> {
+    return this.userRepository.get(userId, withDeleted);
+  }
+
+  async getAdmin(): Promise<UserEntity | null> {
+    return this.userRepository.getAdmin();
+  }
+
+  async getByEmail(email: string, withPassword?: boolean): Promise<UserEntity | null> {
+    return this.userRepository.getByEmail(email, withPassword);
+  }
+
+  async getByOAuthId(oauthId: string): Promise<UserEntity | null> {
+    return this.userRepository.getByOAuthId(oauthId);
+  }
+
+  async getUserProfileImage(user: UserEntity): Promise<ReadStream> {
+    if (!user.profileImagePath) {
+      throw new NotFoundException('User does not have a profile image');
+    }
+    await fs.access(user.profileImagePath, constants.R_OK | constants.W_OK);
```

Comment:
Not sure why we need to check for write access in order to read the profile image.

Reply:
It will crash the server if the file is not there